docs/replication-retry.md

```diff
@@ -0,0 +1,137 @@
+# Replication Retry
+
+## Description
+
+This feature offers a way for users to monitor and retry failed replication
+operations. It enables users to retrieve a list of failed operations and retry
+specific replication operations.
+
+## Requirements
+
+* The listing of failed operations.
+* The retry of a specific failed operation.
+
+## Design
+
+A RESTful API will expose methods for users to list and retry failed operations.
+See [Explanations](#explanations) for more detail.
+
+## Definition of API
+
+* GET `/_/crr/failed/all`
+
+    This GET request retrieves a listing of all failed operations. This
+    operation is useful if you're interested in whether any replication
+    operations have failed and want the entire listing.
+
+    Response:
+
+    ```sh
+    [{
+        bucket: <bucket>,
+        key: <key>,
+        versionId: <versionId>,
+        site: <site>
+    }]
+    ```
+
+* GET `/_/crr/failed/<bucket>/<key>/<versionId>`
+
+    This GET request retrieves a listing of all failed operations for a specific
+    object version. This operation is useful if you're interested in monitoring
+    the replication status of a specific object.
+
+    Response:
+
+    ```sh
+    [{
+        bucket: <bucket>,
+        key: <key>,
+        versionId: <versionId>,
+        site: <site>
+    }]
+    ```
+
+* POST `/_/crr/failed/retry`
+
+    This POST request retries a failed operation.
+
+    Request Body:
+
+    ```sh
+    [{
+        bucket: <bucket>,
+        key: <key>,
+        versionId: <versionId>,
+        site: <site>
+    }]
+    ```
+
+    Response:
+
+    ```sh
+    [{
+        bucket: <bucket>,
+        key: <key>,
+        versionId: <versionId>,
+        site: <site>,
+        status: PENDING
+    }]
+    ```
+
+## Explanations
+
+### Redis
+
+The replication status processor sets a Redis key for any backend with a FAILED
+status with the following key schema:
+
+```sh
+backbeat:crr:failed:<bucket>:<key>:<versionId>:<site>
```

Comment:
Have you evaluated Lists data structure instead of keys? How are you planning on searching the key pattern, using `keys <prefix>*` or `srange`? You may want to do some sizing and performance tests comparing these before choosing the data structure. Putting a large number of keys using test scripts would help.

Reply:
I would like to use a Redis hash where the hash key will be `backbeat:crr:failed`. The hash fields then would be `<bucket>:<key>:<versionId>:<site>`, the field value being the object's metadata. Both `GET` and `HGET` have a constant time complexity O(1), but we would only have one key using a hash which is appealing to me. I am not considering Lists because I want a data structure with a key-value schema.

With a hash, we can store up to 23^2 - 1 field-value pairs, and probably if we reach that limit then we have a greater problem on our hands. Also, as per Redis docs, "small hashes (i.e., a few elements with small values) are encoded in special way in memory that make them very memory efficient". I think this is relevant since the number of failures will likely be small at any given point.

Reply:
For some more numbers using keys with mock object metadata: 
* The serialized length of a hash with 1,000,000 keys is 989887235.
* The serialized length of a string key is 1055 or (1,000,000 * 1055) = 1055000000.